EIPS/eip-5646.md

```diff
@@ -0,0 +1,236 @@
+---
+eip: 5646
+title: Token state fingerprint
+description: Define a way how to distinguish between different states of the same token without implementation details.
+author: Naim Ashhab (@ashhanai)
+discussions-to: https://ethereum-magicians.org/t/eip-5646-discussion-token-state-fingerprint/10808
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-09-11
+requires: 165, 721
+---
+
+## Abstract
+Extend [EIP-721](./eip-721.md) standard with new function `getStateFingerprint`, which returns a unique 32 bytes fingerprint of a token state. This enables distinguishing between different states of the same token and referencing a specific token state without knowing anything about its state structure. It does not define *how* should be state fingerprint computed or stored.
+
+## Motivation
+More and more non-fungible tokens are starting to be "dynamic" in a sense, that the owner can change tokens state, which could affect tokens intrinsic value. A lot of marketplaces have to start taking this into account when managing users offers to buy or accept a particular token as a collateral.
+
+Currently there is no UX friendly way to prevent cases, where a lender makes an offer to a dynamic NFT and borrower withdraws most of its value before accepting the offer. One solution is to use wrapper / bundler, to disable owners' right to modify tokens state, but this brings additional inconveniences. It costs unnecessary gas and users cannot use tokens utility, that doesn't change tokens state (voting rights, NFT gate access, airdrops).
+
+Instead of using wrappers, marketplaces can use state fingerprints as a part of an offer. In that case, an offer that was made to a token in one state, would be invalid for the same token (same token id) in a different state.
+
+Some attempts to standardize token state already existed in the past, for example [Non-Fungible Token State Verification](https://github.com/ethereum/EIPs/issues/1180) and some tokens already implemented similar functions in their contracts, see [MokensDelegate](https://etherscan.io/address/0x8e332fe4df2876cce3b9e6cab160c61ef4c31f77#code)s function `getStateHash(uint256 _tokenId) public view returns (bytes32 stateHash)`. Unfortunately, no EIP standard of this kind has been adopted to this day.
```

Comment:
```suggestion
An increasing number of NFTs are becoming "dynamic," where owners can change tokens' states. A lot of marketplaces have begun taking this into account when managing users' offers to, for example, buy or accept a particular token as collateral.

Currently, there is no application-friendly way to prevent cases where a lender makes an offer to a dynamic NFT and the borrower expends most of the NFT's value before accepting the offer. One solution is to use a wrapper/bundler in order to disable owners' right to modify tokens' states, but this is inconvenient, costing unnecessary gas and preventing the use of tokens in ways that don't change tokens' states (voting rights, NFT gate access, airdrops).

Instead of using wrappers, marketplaces can use state fingerprints as a part of an offer. In that case, an offer that was made to a token in one state would be invalid for the same token in a different state.
```
No external links.